### PR TITLE
bugfix: skip write index.json if no data is wrote.

### LINKED ISSUE #19438: Bug: SteamingDataset Processor data config mismatch
### Bug description

When the number of processes is less than the number of files, some processes do not read or write data. At this time, the data_format written to process_idx_index.json is inconsistent, which leads to the failure of synthesizing the final index.json.

### What version are you seeing the problem on?

v2.2

### How to reproduce the bug

_No response_

### Error messages and logs

image

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by cauyxy**
## What does this PR do?

Skip creating index.json for processes that do not write data.

Fixes #19438 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ yes ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ yes  ] Check that all items from **Before submitting** are resolved
- [ yes ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ yes ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19439.org.readthedocs.build/en/19439/

### CODE REVIEW

**Comment by tchaton** [comment]
Hey @cauyxy Thanks for your contribution!

The index.json file is used to detect whether a given worker has finished its work. This needs to be written down even if the file is empty for the other processes to understand how the workers are done.

This is particularly important when doing multi-node data processing.

**Comment by cauyxy** [comment]
> Hey @cauyxy Thanks for your contribution!
> 
> Would you mind adding a test after verifying the test breaks using master? Especially one emulated multiple nodes. If you get blocked, I can help there.
> 
> BTW, what's your interest in Lightning Data? I am quite curious as you are the first community contributor ;)

Sorry, I'm not very familiar with adding tests, and it might take me some time to do it myself. If you could help with that, it would be great!

I've been experimenting with training some small language models and have used the Lightning framework. For the data part, I primarily utilized StreamingDataset, which I found very convenient and it solved many of my problems. Thank you for your great work.

The code looks very neat, which is extremely pleasing for someone like me who has a strong preference for clean code!🥰

**Comment by tchaton** [comment]
Hey @cauyxy, this is kind. Thank you.

For LLM, we are a couple of Studios I built to display how to best use Lightning Data:
* [Benchmark cloud data-loading libraries](https://lightning.ai/lightning-ai/studios/benchmark-cloud-data-loading-libraries?view=public&section=data+processing)
* [Prepare the TinyLlama 1T token dataset](https://lightning.ai/lightning-ai/studios/prepare-the-tinyllama-1t-token-dataset)
* [Pretrain LLMs - TinyLlama 1.1B](https://lightning.ai/lightning-ai/studios/pretrain-llms-tinyllama-1-1b)

Studios are fully reproducible cloud environments with data, code, dependencies, etc... Give it a spin ;) 

> Sorry, I'm not very familiar with adding tests
This is entirely fine. Do you think you could provide a reproducible dummy python script that triggers the error? This is usually the first step to build a test. Same for the other error ;)

**Comment by cauyxy** [comment]
@tchaton Sure, I can give it a try. I don't have access to Studio right now, but I can set it up locally.

**Comment by cauyxy** [comment]
> Hey @cauyxy, this is kind. Thank you.
> 
> For LLM, we are a couple of Studios I built to display how to best use Lightning Data:
> 
> * [Benchmark cloud data-loading libraries](https://lightning.ai/lightning-ai/studios/benchmark-cloud-data-loading-libraries?view=public&section=data+processing)
> * [Prepare the TinyLlama 1T token dataset](https://lightning.ai/lightning-ai/studios/prepare-the-tinyllama-1t-token-dataset)
> * [Pretrain LLMs - TinyLlama 1.1B](https://lightning.ai/lightning-ai/studios/pretrain-llms-tinyllama-1-1b)
> 
> Studios are fully reproducible cloud environments with data, code, dependencies, etc... Give it a spin ;)
> 
> > Sorry, I'm not very familiar with adding tests
> 
> This is entirely fine. Do you think you could provide a reproducible dummy python script that triggers the error? This is usually the first step to build a test. Same for the other error ;)

this dummy script can trigger pdb.
Env: Linux python3.9


```python
import os
import time
import json
import torch
from pathlib import Path

from lightning.pytorch.utilities.imports import RequirementCache
from lightning.data.streaming import DataChunkRecipe, DataProcessor

assert os.cpu_count() > 1, "cpu count should be greater than 1"
assert RequirementCache("lightning==2.2.0"), "Please use lightning==2.2.0"
assert RequirementCache("lightning_cloud==0.5.64")


base_dir = Path(os.path.dirname(__file__))


def prepare_data():
    data_dir = base_dir / "data"
    data_dir.mkdir(parents=True, exist_ok=True)

    with open(data_dir / "dummy.json", "w") as f:
        f.write(json.dumps({"text": "hello world"}))


prepare_data()


class CommonDataRecipe(DataChunkRecipe):
    def prepare_structure(self, input_dir):
        files = Path(input_dir).rglob("*.json")
        return [str(file) for file in files]

    def prepare_item(self, filepath):
        with open(filepath, "r", encoding="utf-8") as f:
            for _ in f:
                yield torch.tensor([0, 1, 2, 3, 4, 5, 6])


def prepare(
    input_dir: str = "data",
    output_dir: str = "processed",
    chunk_size: int = 4,
) -> None:
    data_recipe = CommonDataRecipe(chunk_size=chunk_size)
    data_processor = DataProcessor(
        input_dir=str(input_dir),
        output_dir=str(output_dir),
        num_workers=os.cpu_count(),
        num_downloaders=1,
    )

    start_time = time.time()
    data_processor.run(data_recipe)
    elapsed_time = time.time() - start_time
    print(f"Time taken: {elapsed_time:.2f} seconds")


if __name__ == "__main__":
    import fire

    fire.Fire(prepare)
```

**Comment by cauyxy** [comment]
> Hey @cauyxy This is great. Now, you can convert this reproducible script into a test ;) You are almost there.

Done!

You are really a good teacher! haha...😂

**Review by tchaton — APPROVED** [review]
Thanks for the test! The breakpoint wasn't meant to be there. So you didn't need to patch builtin. However, if you need to do patching, use monkeypatch, so the patching exists only in the context of the test.

**Comment by cauyxy** [comment]
@tchaton  Happy Chinese New Year! 🍊🐉🏮🧧
